## Paraphrases Template

In [46]:
from google.colab import drive
drive.mount('/content/drive')
BASE_PATH = '/content/drive/My Drive/MAI/Quinto Semestre/IHLT/assets/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [47]:
import pandas as pd

In [48]:
dt = pd.read_csv(f'{BASE_PATH}test-gold/STS.input.SMTeuroparl.txt',sep='\t',header=None)

In [49]:
dt.head(10)

,0,1
0,The leaders have now been given a new chance a...,The leaders benefit aujourd' hui of a new luck...
1,Amendment No 7 proposes certain changes in the...,Amendment No 7 is proposing certain changes in...
2,Let me remind you that our allies include ferv...,I would like to remind you that among our alli...
3,The vote will take place today at 5.30 p.m.,The vote will take place at 5.30pm
4,"The fishermen are inactive, tired and disappoi...","The fishermen are inactive, tired and disappoi..."
5,Neither was there a qualified majority within ...,There was not a majority voting in Parliament ...
6,It increases the power of the big countries at...,It has the effect of augmenting the potency of...
7,"The fishermen are inactive, tired and disappoi...","The fishers are inactive, tired and disappointed."
8,The leaders have now been given a new chance a...,Leaders now have another chance to let them an...
9,The vote will take place today at 5.30 p.m.,The vote will take place at 17h30.


In [50]:
dt['gs'] = pd.read_csv(f'{BASE_PATH}test-gold/STS.gs.SMTeuroparl.txt',sep='\t',header=None)

In [51]:
dt.shape

(459, 3)

In [52]:
dt.head()

,0,1,gs
0,The leaders have now been given a new chance a...,The leaders benefit aujourd' hui of a new luck...,4.50
1,Amendment No 7 proposes certain changes in the...,Amendment No 7 is proposing certain changes in...,5.00
2,Let me remind you that our allies include ferv...,I would like to remind you that among our alli...,4.25
3,The vote will take place today at 5.30 p.m.,The vote will take place at 5.30pm,4.50
4,"The fishermen are inactive, tired and disappoi...","The fishermen are inactive, tired and disappoi...",5.00


## TODO
1. Compute the Jaccard similarity between every paraphrase and add a column *jaccard* to *dt* variable.
2. Compute the pearson correlation as: <br>
```
from scipy.stats import pearsonr
pearsonr(dt['gs'], dt['jaccard'])[0]
```

## NLTK Tokenizer

In [53]:
import nltk
nltk.download('punkt_tab')

dt_tokens = dt.loc[:, [0, 1]].applymap(nltk.word_tokenize)
dt_tokens.head(10)
dt_tokens[0][0]

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
/tmp/ipykernel_1935/1455519710.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  dt_tokens = dt.loc[:, [0, 1]].applymap(nltk.word_tokenize)


['The',
 'leaders',
 'have',
 'now',
 'been',
 'given',
 'a',
 'new',
 'chance',
 'and',
 'let',
 'us',
 'hope',
 'they',
 'seize',
 'it',
 '.']

## Compute the Jaccard similarity

In [54]:
from nltk.metrics.distance import jaccard_distance
dt['jaccard'] = dt_tokens.apply(lambda row: jaccard_distance(set(row[0]), set(row[1])), axis=1)
dt.head(10)

,0,1,gs,jaccard
0,The leaders have now been given a new chance a...,The leaders benefit aujourd' hui of a new luck...,4.500,0.653846
1,Amendment No 7 proposes certain changes in the...,Amendment No 7 is proposing certain changes in...,5.000,0.214286
2,Let me remind you that our allies include ferv...,I would like to remind you that among our alli...,4.250,0.608696
3,The vote will take place today at 5.30 p.m.,The vote will take place at 5.30pm,4.500,0.454545
4,"The fishermen are inactive, tired and disappoi...","The fishermen are inactive, tired and disappoi...",5.000,0.000000
5,Neither was there a qualified majority within ...,There was not a majority voting in Parliament ...,5.000,0.666667
6,It increases the power of the big countries at...,It has the effect of augmenting the potency of...,4.667,0.666667
7,"The fishermen are inactive, tired and disappoi...","The fishers are inactive, tired and disappointed.",5.000,0.200000
8,The leaders have now been given a new chance a...,Leaders now have another chance to let them an...,5.000,0.739130
9,The vote will take place today at 5.30 p.m.,The vote will take place at 17h30.,5.000,0.363636


## Compute the pearson correlation



In [55]:
from scipy.stats import pearsonr
pearsonr(dt['gs'], dt['jaccard'])[0]

np.float64(-0.4504977169318685)

## Justification



We got a negative correlation of about -0.45 because the two scores measure opposite things. The ground truth score (gs) goes up when sentences are very similar, but the Jaccard distance score goes up when sentences are very different. Since high similarity means a low distance, one score goes up while the other goes down, which is why the correlation is negative. The correlation scale goes from -1 to 1. A score of -1 means a perfect opposite relationship, 0 means no relation at all, and 1 means a perfect match. Our result of -0.45 simply means they have a clear, medium-strength opposite relationship.